Error while loading a pretrained EncoderDecoder model

Hi all,
I've trained an `EncoderDecoder` model but I cannot load it via `AutoModelForSeq2SeqLMWithValueHead.from_pretrained`.

My `EncoderDecoder` model has been defined by tying together two RoBERTa model, as follows:

```python
from transformers import EncoderDecoderModel

pretrained_encoder = "seyonec/ChemBERTa-zinc-base-v1"
bert2bert = EncoderDecoderModel.from_encoder_decoder_pretrained(
    pretrained_encoder,
    pretrained_encoder,
    tie_encoder_decoder=True,
)
```

Transformers and TRL versions:

```
trl.__version__: 0.5.0
transformers.__version__: 4.31.0
```

Code to reproduce the issue:

```python
import torch
from transformers import (
    EncoderDecoderConfig,
    AutoTokenizer,
)
from trl import (
    AutoModelForSeq2SeqLMWithValueHead,
    PPOConfig,
    PPOTrainer,
    create_reference_model,
)

pretrained_model = "ribesstefano/ChemBERTa2ChemBERTa-58M"
model = AutoModelForSeq2SeqLMWithValueHead.from_pretrained(pretrained_model)
model_ref = create_reference_model(model, num_shared_layers=6)

tokenizer = AutoTokenizer.from_pretrained(pretrained_model)
tokenizer.pad_token = tokenizer.eos_token
```

Error message:

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
[<ipython-input-45-7d8719dacb7f>](https://localhost:8080/#) in <cell line: 22>()
     20 # print(bert2bert.config.hidden_size)
     21 
---> 22 model = AutoModelForSeq2SeqLMWithValueHead.from_pretrained(pretrained_model)
     23 # # model_ref = AutoModelForSeq2SeqLMWithValueHead.from_pretrained(pretrained_model, device_map="auto", load_in_8bit=True)
     24 # model_ref = create_reference_model(model, num_shared_layers=6)

[/usr/local/lib/python3.10/dist-packages/trl/models/modeling_base.py](https://localhost:8080/#) in from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs)
    230                     raise ValueError("PromptLearningConfig is not supported for PPO training.")
    231         # Then, create the full model by instantiating the wrapper class
--> 232         model = cls(pretrained_model, **trl_model_args)
    233 
    234         # if resume_training, load the state_dict again - this is ok since the

[/usr/local/lib/python3.10/dist-packages/trl/models/modeling_value_head.py](https://localhost:8080/#) in __init__(self, pretrained_model, **kwargs)
    289             raise ValueError("The model does not have a language model head, please use a model that has one.")
    290 
--> 291         self.v_head = ValueHead(self.pretrained_model.config, **v_head_kwargs)
    292 
    293         self._init_weights(**v_head_kwargs)

[/usr/local/lib/python3.10/dist-packages/trl/models/modeling_value_head.py](https://localhost:8080/#) in __init__(self, config, **kwargs)
     37             hidden_size = config.word_embed_proj_dim
     38         else:
---> 39             hidden_size = config.hidden_size
     40 
     41         self.summary = nn.Linear(hidden_size, 1)

[/usr/local/lib/python3.10/dist-packages/transformers/configuration_utils.py](https://localhost:8080/#) in __getattribute__(self, key)
    259         if key != "attribute_map" and key in super().__getattribute__("attribute_map"):
    260             key = super().__getattribute__("attribute_map")[key]
--> 261         return super().__getattribute__(key)
    262 
    263     def __init__(self, **kwargs):

AttributeError: 'EncoderDecoderConfig' object has no attribute 'hidden_size'
```

From what I can guess from the source code, `ValueHead` only inspect the model's top config, expecting an `hidden_size`. However, that information is in the `model.decoder.config.hidden_size` for an `EncoderDecoder` class. Does it sound like a bug or something to update in the code? Or is there any workaround I can try?

Thanks in advance

@younesbelkada I've made a way simpler modification, as you suggested, and just changed the `ValueHead` init in file `trl/models/modeling_value_head.py`:

```python
class ValueHead(nn.Module):
    r"""
    The ValueHead class implements a head for GPT2 that returns a scalar for each output token.
    """

    def __init__(self, config, **kwargs):
        super().__init__()
        if not hasattr(config, "summary_dropout_prob"):
            summary_dropout_prob = kwargs.pop("summary_dropout_prob", 0.1)
        else:
            summary_dropout_prob = config.summary_dropout_prob

        self.dropout = nn.Dropout(summary_dropout_prob) if summary_dropout_prob else nn.Identity()

        # some models such as OPT have a projection layer before the word embeddings - e.g. OPT-350m
        if hasattr(config, "hidden_size"):
            hidden_size = config.hidden_size
        if hasattr(config, "word_embed_proj_dim"):
            hidden_size = config.word_embed_proj_dim
        elif hasattr(config, "is_encoder_decoder"):
            if config.is_encoder_decoder and hasattr(config, "decoder"):
                if hasattr(config.decoder, "hidden_size"):
                    hidden_size = config.decoder.hidden_size

        self.summary = nn.Linear(hidden_size, 1)

        self.flatten = nn.Flatten()
```

After all, the `config`s that the `ValueHead` is expecting are from the pretrained model, so it is sufficient to check those only.

However, when running tests, to which I've added the https://huggingface.co/trl-internal-testing/tiny-Roberta-EncoderDecoderModel model, it breaks for exactly that added `tiny-Roberta-EncoderDecoderModel` model:

```
===================================================================================================================== FAILURES =====================================================================================================================
____________________________________________________________________________________________ Seq2SeqValueHeadModelTester.test_transformers_bf16_kwargs _____________________________________________________________________________________________
[gw0] linux -- Python 3.10.8 /opt/conda/bin/python

self = <tests.test_modeling_value_head.Seq2SeqValueHeadModelTester testMethod=test_transformers_bf16_kwargs>

    def test_transformers_bf16_kwargs(self):
        r"""
        Test if the transformers kwargs are correctly passed
        Here we check that loading a model in half precision works as expected, i.e. the weights of
        the `pretrained_model` attribute is loaded in half precision and you can run a dummy
        forward pass without any issue.
        """
        for model_name in self.all_model_names:
            trl_model = self.trl_model_class.from_pretrained(model_name, torch_dtype=torch.bfloat16)
    
            lm_head_namings = self.trl_model_class.lm_head_namings
    
            if model_name == "trl-internal-testing/tiny-random-FSMTForConditionalGeneration":
                # skip the test for FSMT as it does not support mixed-prec
                continue
    
>           self.assertTrue(
                any(hasattr(trl_model.pretrained_model, lm_head_naming) for lm_head_naming in lm_head_namings)
            )
E           AssertionError: False is not true

tests/test_modeling_value_head.py:430: AssertionError
```

Do you think the error is this related to the modifications I've done in `ValueHead`?

--- a/trl/models/modeling_value_head.py
+++ b/trl/models/modeling_value_head.py
@@ -33,10 +33,14 @@ class ValueHead(nn.Module):
         self.dropout = nn.Dropout(summary_dropout_prob) if summary_dropout_prob else nn.Identity()
 
         # some models such as OPT have a projection layer before the word embeddings - e.g. OPT-350m
+        if hasattr(config, "hidden_size"):
+            hidden_size = config.hidden_size
         if hasattr(config, "word_embed_proj_dim"):
             hidden_size = config.word_embed_proj_dim
-        else:
-            hidden_size = config.hidden_size
+        elif hasattr(config, "is_encoder_decoder"):
+            if config.is_encoder_decoder and hasattr(config, "decoder"):
+                if hasattr(config.decoder, "hidden_size"):
+                    hidden_size = config.decoder.hidden_size
 
         self.summary = nn.Linear(hidden_size, 1)
 

